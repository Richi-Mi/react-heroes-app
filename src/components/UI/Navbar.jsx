import React, { useContext } from 'react';
import { NavLink, useNavigate } from 'react-router-dom';
import AuthContext from '../../auth/authContext';
import types from '../../types/types';

import './../../design/UI/nav.css';

// https://reactrouter.com/docs/en/v6

const Navbar = () => {
    const { user, dispatch } = useContext( AuthContext );
    const navigate = useNavigate();

    const handleLogOut = () => {
        const actionLogOut = {
            type: types.logout
        }
        dispatch( actionLogOut )
        navigate('/login', {
            replace: true // Argumento Replace
        })
    }
    const handleClassLink = ({ isActive }) => isActive ? 'nav_link active' : 'nav_link'
        
    return (
        <nav className='nav'>

            <div className="nav_options">
                <NavLink
                    className={ handleClassLink }
                    to="/"
                >
                    <i className="fas fa-home"></i>
                    Inicio
                </NavLink>
                <NavLink
                    className={ handleClassLink }
                    to="/marvel"
                >
                    Marvel
                </NavLink>

                <NavLink
                    className={ handleClassLink }
                    to="/dc"
                >
                    DC
                </NavLink>
                <NavLink
                    className={ handleClassLink }
                    to="/search"
                >
                    Search
                    <i className="fas fa-search"></i>
                </NavLink>
            </div>
            <div className='nav_acount'>
                <span className='text_info acount_item'>
                    { user.name }
                </span>
                <button
                    onClick={ handleLogOut }
                    className="btn_logout acount_item"
                >
                    Logout
                </button>
            </div>
        </nav>
    )
};

export default Navbar;
