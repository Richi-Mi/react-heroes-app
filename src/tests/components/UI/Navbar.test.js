import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import AuthContext from "../../../auth/authContext";
import Navbar from "../../../components/UI/Navbar";
import types from "../../../types/types";

const mockNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))

describe(' * Pruebas en el componente <Navbar />', () => {

    const name = 'Pedro';
    const mockDispatch = jest.fn();

    const value = {
        user: {
            logged: true,
            name
        },
        dispatch: mockDispatch
    }

    
    const wrapperNav = mount(
        <AuthContext.Provider value={ value }>
            <MemoryRouter initialEntries={ ['/'] }>
                <Navbar />
            </MemoryRouter>
        </AuthContext.Provider>
    );

    beforeEach( () => {
        jest.clearAllMocks()
    })
        
    test(' - Debe de mostrar el componente correctamente', () => {

        expect( wrapperNav ).toMatchSnapshot();
        expect( wrapperNav.find('.text_info').text().trim() ).toBe( name );

    });

    test(' - Debe llamar el logout, dispatch y el navigate con los argumentos requeridos', () => {
        wrapperNav.find('.btn_logout').simulate('click');

        // Llamada del Dispatch
        expect( mockDispatch ).toHaveBeenCalled();
        expect( mockDispatch ).toHaveBeenCalledTimes(1);
        expect( mockDispatch ).toHaveBeenCalledWith({ type: types.logout });

        // Llamada del Navigate
        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledTimes(1);
        expect( mockNavigate ).toHaveBeenCalledWith('/login', { replace: true })
    });


});
