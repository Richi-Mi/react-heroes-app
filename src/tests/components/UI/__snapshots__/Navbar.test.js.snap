// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` * Pruebas en el componente <Navbar />  - Debe de mostrar el componente correctamente 1`] = `
<nav
  className="nav"
>
  <div
    className="nav_options"
  >
    <a
      aria-current="page"
      className="nav_link active"
      href="/"
      onClick={[Function]}
    >
      <i
        className="fas fa-home"
      />
      Inicio
    </a>
    <a
      className="nav_link"
      href="/marvel"
      onClick={[Function]}
    >
      Marvel
    </a>
    <a
      className="nav_link"
      href="/dc"
      onClick={[Function]}
    >
      DC
    </a>
    <a
      className="nav_link"
      href="/search"
      onClick={[Function]}
    >
      Search
      <i
        className="fas fa-search"
      />
    </a>
  </div>
  <div
    className="nav_acount"
  >
    <span
      className="text_info acount_item"
    >
      Pedro
    </span>
    <button
      className="btn_logout acount_item"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</nav>
`;
